x/ibc-hooks/wasm_hook.go

```diff
@@ -0,0 +1,379 @@
+	jsonObject = make(map[string]interface{})
+
+	// If there is no memo, the packet was either sent with an earlier version of IBC, or the memo was
+	// intentionally left blank. Nothing to do here. Ignore the packet and pass it down the stack.
+	if len(memo) == 0 {
+		return false, jsonObject
+	}
+
+	// the jsonObject must be a valid JSON object
+	err := json.Unmarshal([]byte(memo), &jsonObject)
+	if err != nil {
+		return false, jsonObject
+	}
+
+	// If the key doesn't exist, there's nothing to do on this hook. Continue by passing the packet
+	// down the stack
+	_, ok := jsonObject[key]
+	if !ok {
+		return false, jsonObject
+	}
+
+	return true, jsonObject
+}
+
+func ValidateAndParseMemo(memo string, receiver string) (isWasmRouted bool, contractAddr sdk.AccAddress, msgBytes []byte, err error) {
+	isWasmRouted, metadata := jsonStringHasKey(memo, "wasm")
+	if !isWasmRouted {
+		return isWasmRouted, sdk.AccAddress{}, nil, nil
+	}
+
+	wasmRaw := metadata["wasm"]
+
+	// Make sure the wasm key is a map. If it isn't, ignore this packet
+	wasm, ok := wasmRaw.(map[string]interface{})
+	if !ok {
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, "wasm metadata is not a valid JSON map object")
+	}
+
+	// Get the contract
+	contract, ok := wasm["contract"].(string)
+	if !ok {
+		// The tokens will be returned
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, `Could not find key wasm["contract"]`)
+	}
+
+	contractAddr, err = sdk.AccAddressFromBech32(contract)
+	if err != nil {
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, `wasm["contract"] is not a valid bech32 address`)
+	}
+
+	// The contract and the receiver should be the same for the packet to be valid
+	if contract != receiver {
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, `wasm["contract"] should be the same as the receiver of the packet`)
+	}
+
+	// Ensure the message key is provided
+	if wasm["msg"] == nil {
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, `Could not find key wasm["msg"]`)
+	}
+
+	// Make sure the msg key is a map. If it isn't, return an error
+	_, ok = wasm["msg"].(map[string]interface{})
+	if !ok {
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, `wasm["msg"] is not a map object`)
+	}
+
+	// Get the message string by serializing the map
+	msgBytes, err = json.Marshal(wasm["msg"])
+	if err != nil {
+		// The tokens will be returned
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, err.Error())
+	}
+
+	return isWasmRouted, contractAddr, msgBytes, nil
+}
+
+func (h WasmHooks) SendPacketOverride(i ICS4Middleware, ctx sdk.Context, chanCap *capabilitytypes.Capability, packet ibcexported.PacketI) error {
```

Comment:
This implements this ibc_callback feature. Do we need it?
It works like this: we send message with memo: {"ibc_callback": "ntrncontractaddress"},
and will get a sudo callback on this address later.
We already kinda have same functionality, right?

Reply:
Let's disable callbacks from hooks on Neutron side.
So if there is a transfer from Neutron to some other chain, a callback will happen through our system.